### ISSUE TITLE
JWT auth system with double cookies

### ISSUE THREAD

**Issue by ArchOrn** [ktorio/ktor#1297]
### Ktor Version and Engine Used (client or server and name)

Ktor v1.2.3 (Netty Engine)
Kotlin v1.3.50 (JDK 8)

### Issue description

Hi,

I am currently trying to implement an auth system for my API using JWT with headers. I succeeded using the `fun Route.authenticate { ... }` and passing a proper `Authorization` header, but I do not have currently on my Vue.js app a persistence for my JWT token as I read that it is not really a good practice (there is a lot of article about that saying everything and its contrary, so I am a bit lost...).
I found this [article](https://medium.com/lightrail/getting-token-authentication-right-in-a-stateless-single-page-application-57d0c6474e3) which explains how a system we can setup that using double cookie to avoid storing the whole JWT token.

My problem is achieving to reassemble the whole token using the `call.request.cookies` information and the `fun authHeader(block: (ApplicationCall) -> HttpAuthHeader?) { ... }` function that allows to change the way the token is retrieved. I use this inside the `install(Authentication) { jwt { ... } }` configuration block, but it seems that the cookies are not set at that time, whereas inside my login route `route.post("/login") { ... }`, they are defined.
While debugging I can see that the route call happens before the authentication method so I do not understand why the cookies are not available. Is it a bug or is there something I do not understand about `JWTAuthenticationProvider.Configuration` scope?

Here is an example of what I am trying to achieve:

```kotlin
private fun JWTAuthenticationProvider.Configuration.jwtAuthConfiguration()
{
  // Check JWT parts in cookies and alter header if any
  authHeader { call ->
    val oldHeader = call.request.parseAuthorizationHeaderOrNull()
    val headerPayload = call.request.cookies[JWT_HEADER_PAYLOAD_NAME]
    val signature = call.request.cookies[JWT_SIGNATURE_NAME]
    headerPayload?.let { first -> signature?.let { second -> "$first.$second" } }?.let { jwt ->
      HttpAuthHeader.Single(oldHeader?.authScheme ?: "Bearer", jwt)
    } ?: oldHeader
    // TODO: Check CSRF
  }

  realm = jwtRealm
  verifier(makeJwtVerifier(jwtAudience, jwtIssuer))
}
```

Thank you for your support.
Regards,

Raphaël

**Comment by cy6erGn0m** [comment]
Request cookies are always available during a call processing despite phase and computed lazily on-demand. Are you sure that cookies are not defined? Can't it be a debugger's glitch?

**Comment by ArchOrn** [comment]
Ok, after a bit of checking, with logs and such, as you mentioned that it could be a debugger issue, I started to realize that it was strange that only the login request was blessed with some cookies...
And in fact, the problem is, as per the [RFC-6265](https://tools.ietf.org/html/rfc6265#section-5.2.4), when you do not provide any `path` parameter to the `RequestCookies` items, it defaults to the current path, i. e. `/api/v1/auth` in my case, which allows only the `/login` and the `/register` requests to send cookies, just where I was not needing them.

Sorry for the auto responding issue, I set `Path=/api` to allow cookies to be sent within every API requests. Hope this will help :)
